modules/core/src/test/java/org/apache/ignite/internal/client/thin/ServiceAwarenessTest.java

```diff
@@ -0,0 +1,624 @@
+                svc.testMethod();
+
+            ((GridTestLog4jLogger)log).setLevel(Level.INFO);
+
+            assertTrue(srvcTopOnClient.size() == INIT_SRVC_NODES_CNT
+                && srvcTopOnClient.contains(grid(1).localNode().id())
+                && srvcTopOnClient.contains(grid(2).localNode().id()));
+
+            long prevTopVersion = grid(0).context().discovery().topologyVersion();
+
+            grid(1).services().cancel(SRV_NAME);
+
+            srvcTopOnClient.clear();
+
+            grid(1).services().deploy(serviceCfg().setNodeFilter(null));
+
+            assertEquals(prevTopVersion, grid(0).context().discovery().topologyVersion());
+
+            ((GridTestLog4jLogger)log).setLevel(Level.DEBUG);
+
+            while (srvcTopOnClient.isEmpty())
+                svc.testMethod();
+
+            assertEquals(GRIDS, srvcTopOnClient.size());
+
+            for (Ignite ig : G.allGrids())
+                assertTrue(srvcTopOnClient.contains(ig.cluster().localNode().id()));
+        }
+    }
+
+    /** */
+    private void doTestClusterTopChangesWhileServiceCalling(
+        int nodesCnt,
+        boolean addNodes,
+        boolean multiThreaded)
+        throws Exception {
+        assert nodesCnt > 0;
+
+        Set<UUID> newNodesUUIDs = new GridConcurrentHashSet<>();
+
+        // Start additional nodes to stop them.
+        if (!addNodes) {
+            startGridsMultiThreaded(GRIDS, nodesCnt);
+
+            for (int i = GRIDS; i < GRIDS + nodesCnt; ++i)
+                newNodesUUIDs.add(grid(i).localNode().id());
+        }
+
+        // Service topology on the clients.
+        Set<UUID> srvcTopOnClient = new GridConcurrentHashSet<>();
+
+        addSrvcTopUpdateClientLogLsnr(srvcTopOnClient::addAll);
+
+        AtomicBoolean changeClusterTop = new AtomicBoolean();
+        AtomicBoolean stopFlag = new AtomicBoolean();
+
+        try (IgniteClient client = startClient()) {
+            ServicesTest.TestServiceInterface svc = client.services().serviceProxy(SRV_NAME, ServicesTest.TestServiceInterface.class);
+
+            ((GridTestLog4jLogger)log).setLevel(Level.DEBUG);
+
+            IgniteInternalFuture<?> runFut = runMultiThreadedAsync(() -> {
+                do {
+                    try {
+                        svc.testMethod();
+                    }
+                    catch (ClientException e) {
+                        String m = e.getMessage();
+
+                        // Until the service topology is not updated yet, service invoke request can be redirected to
+                        // node which has just left the cluster. This case raises a node-left exception in the service
+                        // call response. Unfortunately, this exception is not processed by the client service proxy as
+                        // a resend attempts.
+                        if (addNodes || (!m.contains("Node has left grid") && !m.contains("Failed to send job due to node failure"))
```

Comment:
Not sure I understand the issue here. This exception occurs on the server, right? And the server should do the retry?

Reply:
It is definitely out of scope, but we can create a ticket and mention it with a TODO here.

Reply:
I've put TODO to https://issues.apache.org/jira/browse/IGNITE-20802